### PR TITLE
Wednesday Check up

### PR DESCRIPTION
**Pull request by mikeSheets**

### CODE REVIEW

File: .idea/.generators [hunk 1]
```diff
@@ -5,4 +5,4 @@ You are allowed to:
 2. Remove generators
```
Comment (fiestacasey) [inline .idea/.generators:5]: git ignore this whole file

**Comment by fiestacasey** [comment]
@ksolkowski @kahlya

File: app/assets/javascripts/admin/application.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+// This is a manifest file that'll be compiled into application.js, which will include all the files
```
Comment (fiestacasey) [inline app/assets/javascripts/admin/application.js:1]: convert this to coffeescript

File: app/assets/javascripts/application.js [hunk 3]
```diff
@@ -13,4 +13,5 @@
 //= require jquery
```
Comment (fiestacasey) [inline app/assets/javascripts/application.js:13]: convert to coffescript

File: app/assets/javascripts/cart.coffee [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (fiestacasey) [inline app/assets/javascripts/cart.coffee:1]: delete empty file

File: app/assets/javascripts/home.coffee [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (fiestacasey) [inline app/assets/javascripts/home.coffee:1]: delete empty file (not gonna comment on the rest of the coffeescript ones but they should all go)

File: app/assets/stylesheets/admin/orders.scss [hunk 6]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Orders controller here.
```
Comment (fiestacasey) [inline app/assets/stylesheets/admin/orders.scss:1]: delete empty file

File: app/assets/stylesheets/cart.scss [hunk 7]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Cart controller here.
```
Comment (fiestacasey) [inline app/assets/stylesheets/cart.scss:1]: delete empty file (not gonna comment on the rest of the scss ones but they should all go)

File: app/helpers/cart_helper.rb [hunk 8]
```diff
@@ -0,0 +1,2 @@
+module CartHelper
```
Comment (fiestacasey) [inline app/helpers/cart_helper.rb:1]: delete empty helper files

File: app/models/address.rb [hunk 9]
```diff
@@ -0,0 +1,6 @@
+class Address < ActiveRecord::Base
+	belongs_to :state
```
Comment (fiestacasey) [inline app/models/address.rb:2]: tabs should be spaces in this whole file

File: app/models/credit_card.rb [hunk 10]
```diff
@@ -0,0 +1,25 @@
+class CreditCard < ActiveRecord::Base
+	belongs_to :user
```
Comment (fiestacasey) [inline app/models/credit_card.rb:2]: tabbing

File: app/models/credit_card.rb [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+	has_many :payments, inverse_of: :credit_card
+  validates_presence_of :last_four, :month, :year, :user, :name
+  validates_presence_of :number, :cvc,
+                        on: :create
+  attr_accessor :number, :cvc
+
+  before_validation :set_last_four
+
+  def expiration_date
+    DateTime.parse("#{self.year}-#{self.month}-01").end_of_month
```
Comment (fiestacasey) [inline app/models/credit_card.rb:12]: you don't need `self.` for accessors

File: .gitignore [hunk 12]
```diff
@@ -0,0 +1,10 @@
+
+.idea/IceCreamTreeStore.iml
+
+.idea/IceCreamTreeStore.iml
+
+.idea/workspace.xml
+
```
Comment (ksolkowski) [inline .gitignore:7]: probably a good idea to gitignore everything in .idea

File: app/models/credit_card.rb [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+  end
+
+  private
+  def set_last_four
+    return if self.number.blank?
+    if CreditCardValidator::Validator.valid?(self.number)
```
Comment (fiestacasey) [inline app/models/credit_card.rb:18]: you don't need `self.number` in here

File: app/models/order.rb [hunk 14]
```diff
@@ -0,0 +1,41 @@
+class Order < ActiveRecord::Base
+	belongs_to :user, inverse_of: :orders
```
Comment (fiestacasey) [inline app/models/order.rb:2]: tabbing in this whole file

File: app/models/payment.rb [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+	STATES = [PENDING, COMPLETED, VOIDED]
+	validates(:state, inclusion: STATES)
+
+	before_save :validate_state_change
+
+  after_initialize :set_default_state
+
+	def validate_state_change
+		old,new = changed_attributes["state"]
+		new = state
```
Comment (fiestacasey) [inline app/models/payment.rb:19]: you can't use new, its a reserved word

File: app/views/admin/products/index.json.jbuilder [hunk 16]
```diff
@@ -0,0 +1,4 @@
+json.array!(@products) do |product|
```
Comment (fiestacasey) [inline app/views/admin/products/index.json.jbuilder:1]: huh. whats this?

File: app/views/admin/products/show.html.haml [hunk 17]
```diff
@@ -0,0 +1,34 @@
+%p#notice= notice
+.container
+  %p
```
Comment (fiestacasey) [inline app/views/admin/products/show.html.haml:3]: I don't think you should put a p in a p

File: app/models/order.rb [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+	validates(:state, inclusion: STATES)
+
+	before_save :validate_state_change
+
+	after_initialize :set_default_state
+
+
+	def validate_state_change
+		old = changed_attributes["state"]
+    new = state
```
Comment (ksolkowski) [inline app/models/order.rb:21]: tabbing

File: app/models/payment.rb [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+	PENDING = "pending"
+	COMPLETED = "completed"
+	VOIDED = "voided"
+
+	STATES = [PENDING, COMPLETED, VOIDED]
+	validates(:state, inclusion: STATES)
+
+	before_save :validate_state_change
+
+  after_initialize :set_default_state
```
Comment (ksolkowski) [inline app/models/payment.rb:15]: tabbing

File: app/views/products/index.json.jbuilder [hunk 20]
```diff
@@ -0,0 +1,4 @@
+json.array!(@products) do |product|
```
Comment (fiestacasey) [inline app/views/products/index.json.jbuilder:1]: wut. huh?

File: .gitignore (hunk 12 above)
Comment (fiestacasey) [inline .gitignore:7]: ignore log/\* too

File: test/factories.rb [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+    end
+    sequence :city do |n|
+      "city-#{n}"
+    end
+    state { State.first || create(:state) }
+    zip 00000
+    user { create(:user) }
+  end
+
+	factory :state do
```
Comment (fiestacasey) [inline test/factories.rb:55]: tabbing

File: test/fixtures/addresses.yml [hunk 22]
```diff
@@ -0,0 +1,13 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
```
Comment (fiestacasey) [inline test/fixtures/addresses.yml:2]: delete all fixtures/*

File: test/models/address_test.rb [hunk 23]
```diff
@@ -0,0 +1,12 @@
+require 'test_helper'
+
+class AddressTest < ActiveSupport::TestCase
+  test "valid create" do
+		address = create(:address)
+    assert address.persisted?
+	end
+
+	test "validations" do
```
Comment (fiestacasey) [inline test/models/address_test.rb:9]: tabbing

File: test/models/order_item_test.rb [hunk 24]
```diff
@@ -0,0 +1,12 @@
+require 'test_helper'
+
+class OrderItemTest < ActiveSupport::TestCase
+  test "valid create" do
+		order_item = create(:order_item)
+    assert order_item.persisted?
+	end
+
+	test "validations" do
```
Comment (fiestacasey) [inline test/models/order_item_test.rb:9]: tabbing

File: .gitignore [hunk 25]
```diff
@@ -0,0 +1,10 @@
...
+log/test.log
+
```
Comment (ksolkowski) [inline .gitignore:9]: add `/tmp/*`

File: test/models/order_test.rb [hunk 26]
```diff
@@ -0,0 +1,22 @@
+require 'test_helper'
+
+class OrderTest < ActiveSupport::TestCase
+  test "valid create" do
+    order = create(:order)
+    assert order.persisted?
+  end
+
+  # test "can't transition from cart to canceled" do
```
Comment (fiestacasey) [inline test/models/order_test.rb:9]: commented code should generally be removed. You can always find it later in an old commit

File: test/models/payment_test.rb [hunk 27]
```diff
@@ -0,0 +1,26 @@
+require 'test_helper'
+
+class PaymentTest < ActiveSupport::TestCase
+  test "valid create" do
+		payment = create(:payment)
+    assert payment.persisted?
+	end
+
+	test "validations" do
+		assert_validations(Payment.new, [:amount, :order, :credit_card])
```
Comment (fiestacasey) [inline test/models/payment_test.rb:10]: tabbing

File: test/models/product_test.rb [hunk 28]
```diff
@@ -0,0 +1,14 @@
+require 'test_helper'
+
+class ProductTest < ActiveSupport::TestCase
+
+  test "valid create" do
+
+		product = create(:product)
```
Comment (fiestacasey) [inline test/models/product_test.rb:7]: tabbing

File: test/models/state_test.rb [hunk 29]
```diff
@@ -0,0 +1,15 @@
+require 'test_helper'
+
+class StatesTest < ActiveSupport::TestCase
+  test "valid create" do
+		state = create(:state)
+    assert state.persisted?
+	end
+
+	test "validations" do
```
Comment (fiestacasey) [inline test/models/state_test.rb:9]: tabbing

File: test/models/user_test.rb [hunk 30]
```diff
@@ -0,0 +1,17 @@
+require 'test_helper'
+
+class UserTest < ActiveSupport::TestCase
+  test "valid create" do
+		user = create(:user)
```
Comment (fiestacasey) [inline test/models/user_test.rb:5]: tabbing

File: tmp/cache/assets/sprockets/v3.0/-6/-6UQksjmetglcJbh4NVvu13Gh2CiVb7diTfN1ol1yVU.cache [hunk 31]
```diff
@@ -0,0 +1 @@
+I"�app/assets/stylesheets/application.css.scss?type=text/css&pipeline=debug&id=9b2a47970c5140d808b8743c67e072595f63c5f2d395a7979a1762eb4fce7d0d:ET
```
Comment (fiestacasey) [inline tmp/cache/assets/sprockets/v3.0/-6/-6UQksjmetglcJbh4NVvu13Gh2CiVb7diTfN1ol1yVU.cache:1]: gitignore tmp/\*

File: app/models/order.rb [hunk 32]
```diff
@@ -0,0 +1,41 @@
...
+		old_index = STATES.index(old)
+		new_index = STATES.index(new)
+
+
+		if !persisted?
+			errors.add(:state, "initial state must be #{STATES.first}") if self.state != STATES.first
+		elsif new_index.nil?
+			errors.add(:state, "#{new} is not a valid state")
+		elsif (new_index) != (old_index + 1)
```
Comment (austinkahly) [inline app/models/order.rb:30]: you shouldn't need the parenthesis around new_index

**Pull request merged**